java/core/src/main/java/sleeper/core/iterator/AggregationFilteringIterator.java

```diff
@@ -0,0 +1,483 @@
+            } else {
+                throw new IllegalArgumentException("Value type not implemented " + lhs.getClass());
+            }
+        }
+    }
+
+    /**
+     * Defines an aggregation operation with the column name to perform on it.
+     *
+     * @param column the column to aggregate
+     * @param op     the aggregation operator
+     */
+    public record Aggregation(String column, AggregationOp op) {
+        public Aggregation {
+            Objects.requireNonNull(column, "column");
+        }
+    }
+
+    /**
+     * Defines the filtering and aggregation configuration for this iterator.
+     *
+     * @param  groupingColumns           the columns that the aggregation will "group by"
+     * @param  ageOffColumn              the optional column to have an age-off filter applied
+     * @param  maxAge                    the maximum age in seconds for rows
+     * @param  aggregations              the list of aggregations to apply
+     * @throws IllegalArgumentExpception if {@code groupingColumns} is empty
+     */
+    public record FilterAggregationConfig(List<String> groupingColumns, Optional<String> ageOffColumn, long maxAge, List<Aggregation> aggregations) {
+
+        public FilterAggregationConfig {
+            Objects.requireNonNull(groupingColumns, "groupingColumns");
+            if (groupingColumns.isEmpty()) {
+                throw new IllegalArgumentException("must have at least one grouping column");
+            }
+            Objects.requireNonNull(ageOffColumn, "ageOffColumn");
+            Objects.requireNonNull(aggregations, "aggregations");
+        }
+    }
+
+    @Override
+    public CloseableIterator<Record> apply(CloseableIterator<Record> source) {
+        if (config == null) {
+            throw new IllegalStateException("AggregatingIterator has not been initialised, call init()");
+        }
+        // See if we need to age-off data
+        CloseableIterator<Record> input = maybeCreateFilter(source);
+
+        // Do any aggregations need to be performed?
+        if (!config.aggregations().isEmpty()) {
+            return new AggregatorIteratorImpl(config, input);
+        } else {
+            return input;
+        }
+    }
+
+    /**
+     * Configures an age off filter on the source iterator if needed.
+     *
+     * If this aggregating iterator has been configured to provide age off filtering,
+     * then an {@link AgeOffIterator} is appended to the input iterator to provide
+     * age-off functionality.
+     *
+     * @param  source the input iterator
+     * @return        the source iterator or a new filtering iterator
+     */
+    private CloseableIterator<Record> maybeCreateFilter(CloseableIterator<Record> source) {
+        return config.ageOffColumn().map(filter_col -> {
+            AgeOffIterator ageoff = new AgeOffIterator();
+            // Age off iterator operates in milliseconds
+            ageoff.init(String.format("%s,%d", filter_col, config.maxAge()), schema);
```

Comment:
We need to decide whether this should operate in seconds or milliseconds. AgeOffIterator was only written as an example, I don't think there's a particular reason it uses milliseconds.